io-engine/src/bdev/nexus/nexus_channel.rs

```diff
@@ -8,13 +8,17 @@ use std::{
 
 use super::{ChildState, Nexus, Reason};
 
-use crate::core::{BlockDeviceHandle, CoreError, Cores};
+use crate::{
+    core::{BlockDeviceHandle, CoreError, Cores},
+    rebuild::RebuildLogHandle,
+};
 
 /// io channel, per core
 #[repr(C)]
 pub struct NexusChannel<'n> {
     writers: Vec<Box<dyn BlockDeviceHandle>>,
     readers: Vec<Box<dyn BlockDeviceHandle>>,
+    rebuild_logs: Vec<RebuildLogHandle>,
```

Comment:
We don't have that many children (1-3 mostly), so having a hash map or btree won't have any benefit. From calculation and memory allocation and access standpoint, HashMap and BTree would actually be worse here. For long-running applications, it's better to avoid having non-local data structures like linked list or trees (that's why Rust encourages Vec data structure in many cases).

Another problem with HashMap (as opposed to BTree) is that it doesn't not guarantee item order, which can change from run to run. That we'd like to avoid in I/O path.

Reply:
ok. What does "non-local data structures" mean here? Something in Rust parlance?

Reply:
No, this term is sometimes used to refer to how a data structure is stored memory.

For example, a Vec<> or C++ std::vector<>, or C/C++ array allocate a large, continuous block of memory, storing all element one after another. When you iterate items, CPU reads from roughly the same region, thus with very high number of cache hits. Your next items is very likely to be already prefetched in CPU cache.

A B-tree or linked list allocate a small chuck of memory for each node (data item + pointers to other nodes). When you iterate such structure, CPU has to read from different regions of memory. This has two sonquences:
1) Each element must be accessed by a pointer dereference, which isn't quite friendly to CPU cache.
2) Every next item may lie in a different page, unpredictable by caching CPU logic. This pollutes the CPU cache with data you won't access very soon.
These two things lead to situation when CPU cache may become very ineffective, and performance of a CPU-bound program deteriorates.

That's one of the main contributing factors of relatively poor performance of GC languages like Java. Not the interpreted byte-code as some believe (it is actually compiled of-the-fly into CPU instructions), but the fact there's no memory-local data structures in Java. ArrayList is a vector of points to items, each allocate in a different memory location.

Reply:
Alright, we are talking about 'locality of reference' here :)